src/node/historical_queries_adapter.cpp

```diff
@@ -153,6 +205,49 @@ namespace ccf
 
     return receipt;
   }
+
+  std::vector<uint8_t> describe_merkle_proof_v1(const TxReceiptImpl& receipt)
+  {
+    constexpr size_t buf_size = 2048;
+    Q_USEFUL_BUF_MAKE_STACK_UB(buffer, buf_size);
+
+    QCBOREncodeContext ctx;
+    QCBOREncode_Init(&ctx, buffer);
+
+    QCBOREncode_BstrWrap(&ctx);
+    QCBOREncode_OpenMap(&ctx);
+
+    if (!receipt.commit_evidence)
+    {
+      throw std::logic_error("Merkle proof is missing commit evidence");
+    }
+    if (!receipt.write_set_digest)
+    {
+      throw std::logic_error("Merkle proof is missing write set digest");
+    }
+    encode_leaf_cbor(ctx, receipt);
+
+    if (!receipt.path)
+    {
+      throw std::logic_error("Merkle proof is missing path");
+    }
+    encode_path_cbor(ctx, *receipt.path);
+
+    QCBOREncode_CloseMap(&ctx);
+    QCBOREncode_CloseBstrWrap2(&ctx, false, nullptr);
+
+    struct q_useful_buf_c result;
+    auto qerr = QCBOREncode_Finish(&ctx, &result);
+    if (qerr)
+    {
+      throw std::logic_error(
+        fmt::format("Failed to encode merkle proof: {}", qerr));
+    }
+
+    return {
```

Comment:
Check out the possibility to re-use buffer memory

Reply:
They're actually same, both for `cose_sign1` and here for encode/decode.

But now I think, for `cose_sign1()` you'll have to move buffer size estimation outside, and same for verification, if you want to create buffer in place and pass it inside.

@achamayou  I think it's OK to leave it as is and only copy-paste the decode/encode part, looks simpler.

Reply:
Talked offline, we can create a vector in place instead of a stack buffer

Reply:
Done: https://github.com/microsoft/CCF/pull/6480/commits/0a2105cfc86f8c904b1c1c34a661ae0a43b455bd